### PR TITLE
Add Selector size variations and switch SelectorGroup to horizontal orientation

### PR DESCRIPTION
**Pull request by mykolaharmash**
Addresses  #931.

## Purpose

Extending `Selector` and `SelectorGroup` features to support most common usecases.

## Approach and changes

* Add `size` prop to `Selector`
* Refactor styling of the `Selector` to make sure its height matches other controls like `Input` and `Select`
* Switch `SelectorGroup` to a horizontal orientation on a big screen and vertical on mobile
* Add `orientation` prop to `SelectorGroup` to choose between horizontal and vertical layout.
* Add `size` prop to `SelectorGroup` which changes size of a `Selector` within the group
* Add `gapSize` prop to `SelectorGroup` to control the spacing between `Selectors`
* Add `stretch` prop to `SelectorGroup` to make it fill all available width

## Definition of done

* [x] Development completed
* [x] Reviewers assigned
* [ ] ~~Unit and integration tests~~ Only updated snapshots as it's only styling changes, let me know if you think we need some new tests
* [x] Meets minimum browser support
* [x] Meets accessibility requirements

### CODE REVIEW

File: packages/circuit-ui/components/Selector/Selector.docs.mdx
Comment (robinmetral) [inline packages/circuit-ui/components/Selector/Selector.docs.mdx:40]:
```suggestion
- **flexible**, used for more complex content within selector which may define its own margins. **flexible** just adds a minimal equal margin on all sides.
```

File: packages/circuit-ui/components/Selector/Selector.stories.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/Selector/Selector.stories.tsx:65]: This will need to be `<Body />` in V3. Not sure how we should handle this since the typography PR (https://github.com/sumup-oss/circuit-ui/pull/884) was already merged 🤔 Thoughts @connor-baer?

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/Selector/Selector.tsx:170]: Thanks a lot for adding a comment on what this does! 

Is there a reason we didn't keep the previous implementation? I see this one resorts to `calc()` which might be slightly less performant and looks a bit less clean IMO. But is it doing anything new?

```suggestion
    // Serves only as a container for the border,
```

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/Selector/Selector.tsx:145]: I think this would override the label in baseStyles, I would just declare it there 🙂 (baseStyles are receiving the size as well, if we want to add it to the label)

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:75]: Not sure we should expose these two 🤔 it might be better to let consumers add custom styles for edge cases. What do you think?

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:152]: Why do we need this now? Because of the inline-block instead of block display property of the Selector? If we need to do this here, it means that someone who has implemented the `Selector` by itself in some settings might need a similar fix

**Review by robinmetral — CHANGES_REQUESTED** [review]
Thanks for this great improvement to the Selector/SelectorGroup! 🙌 

My main concern here is with versioning: as I mentioned in some of my comments, this looks like a major. If you need this earlier than Circuit V3, I think it could still be possible but we would need to avoid any breaking changes. I'd also love input from @connor-baer here 🙂 

(Also, sorry I missed the issue at #931, we could have talked about the versioning questions earlier. Generally though, this is definitely the direction we're going into since there are design specs for it, so thanks for opening this PR! Let's figure out how to best release this.)

File: .changeset/hot-cheetahs-press.md
Comment (mykolaharmash) [reply .changeset/hot-cheetahs-press.md:2]: Right, it's a breaking change indeed... How about I add an `orientation` prop (vertical by default)? This would be backward compatible. This would clash with `inlineMobile` but, actually, might be better to get rid of it altogether. Wdyt?

File: packages/circuit-ui/components/Selector/Selector.stories.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/Selector/Selector.stories.tsx:65]: It should be fine if I make this whole change non-breaking, right?

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/Selector/Selector.tsx:170]: Yeah, this one is tricky to explain with text, please see the video https://cln.sh/DpVKQx

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/Selector/Selector.tsx:145]: True, will do 👍

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:75]: At first, I thought exactly that as well, to leave it up to the user. But the issue is that `Selector` needs to be modified as well in order to be able to stretch (see `StyledSelector`), so it does not really make sense to keep in mind this stretching usecase by adjusting internal elements but to not expose the top-level property.

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:152]: It's here to support the `stretch` prop. `Selector` needs to fill the width of `OptionItem` which has `flex-grow: 1;`, otherwise `OptionItem` will stretch but `Selector` within won't.

File: .changeset/hot-cheetahs-press.md
Comment (mykolaharmash) [reply .changeset/hot-cheetahs-press.md:2]: So, I got rid of `inlineMobile` prop and added `orientation` which is `vertical` by default, which is matching the current behavior. `stretch` is now `true` for the same also to match how it's rendered right now. With this, it should be a non-breaking change.

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/Selector/Selector.tsx]: I refactored this part, because I realized that `Input` is not actually using `border`, it mimics it with `box-shadow`, meaning `Selector` and `Input` would still be misaligned when using my previous implementation 🤦‍♂️

What I do now:
* Using `box-shadow` to mimic border, just as the input does
* Using `::before` only as a container for the focus outline (still need to align it with the label's border using `calc()` so that focus outline does not cover label's border)

Let me know if this is not clear, we can have a chat and I'll try to explain things more visually.

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/Selector/Selector.tsx:173]: Realized that it's actually should override the parent label as the one for size is more specific and includes the `selector` part.

File: packages/circuit-ui/components/Selector/Selector.stories.tsx
Comment (robinmetral) [reply packages/circuit-ui/components/Selector/Selector.stories.tsx:65]: Yes it's fine 🙁 We'll go over the codebase again before v3 and address any leftover typography changes

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/Selector/Selector.tsx:87]: Let's do this with multiple comma-separated box-shadows, it will be cleaner 🙂 I managed to get it to work locally on your branch, I'll show you my thinking on our call today

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (robinmetral) [reply packages/circuit-ui/components/Selector/Selector.tsx:173]: Sounds good, then maybe we should make it `selector__label--${size}` for consistency

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.spec.tsx
Comment (robinmetral) [inline packages/circuit-ui/components/SelectorGroup/SelectorGroup.spec.tsx:90]: I think you forgot to pass the orientation prop here 🙂

File: .changeset/hot-cheetahs-press.md
Comment (robinmetral) [inline .changeset/hot-cheetahs-press.md:11]: Let's also update this to reflect the most recent changes 🙂

**Review by robinmetral — COMMENTED** [review]
Thank you so much for addressing the changes! Just a few extra comments here, let's also go through them on our call and wrap this up today 🎉

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.spec.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/SelectorGroup/SelectorGroup.spec.tsx:90]: oops 😬

**Comment by mykolaharmash** [comment]
As agreed, we're going to make it a breaking change and release it with the next major version.

Latest changes:
* Changing the base branch to `next`
* Using `box-shadow` composition to render border and focus outline
* Handling all the `box-shadow` styling in a dedicated function. This makes it easier to follow all the outline-related logic
* Adding a state to track if `Selector` has focus. This also helps consolidate styling for the Label and get rid of `input + label` styles which were in a separate function before.

**Review by robinmetral — COMMENTED** [review]
Just a few last things, it looks great 🎉 Thanks a lot for your work on this

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/Selector/Selector.tsx:83]: Nice, did not notice it 👍

File: .changeset/hot-cheetahs-press.md
Comment (mykolaharmash) [reply .changeset/hot-cheetahs-press.md:2]: You're, it's a mojor, as agreed previous, I just forgot to update this file 🙂

About the defaults, I think it makes sense to change them now to cover more common scenarious. Will do.

**Comment by mykolaharmash** [comment]
Please give it another look, I've addressed the latest comments.

**Review by robinmetral — APPROVED** [review]
Amazing! 🚀 

You simply have outdated snapshots, the rest looks great. Thanks a lot for this!

**Comment by connor-baer** [comment]
Thank you for the contribution @mykolaharmash and the thorough review @robinmetral. I'll take a quick look later today :)

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (connor-baer) [inline packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:70]: Can you share a use case where the gap size would need to be different than the default?

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (connor-baer) [inline packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:62]: I don't think this should be configurable with a prop. The orientation can often change between viewports (vertical on mobile, horizontal on desktop) and it's too complicated to codify this in a prop. 

Circuit UI should provide a sensible default that can be customized with the `css` prop.

**Review by connor-baer — CHANGES_REQUESTED** [review]

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (connor-baer) [reply packages/circuit-ui/components/Selector/Selector.tsx]: The reason why the outline was applied with a pseudo-element was to prevent the selector height from changing when it's checked (and the border increases to 2px).

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (connor-baer) [reply packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:62]: I caught up on the discussions between @robinmetral and you above. I think it's fine to change the default in v3, but I still wouldn't make it configurable.

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/Selector/Selector.tsx:240]: Main reason is to be able to consolidate all the outline-related logic into a single function, so it's easier to follow as it's not straightforward. Without this flag in state focus outline had to live in the styles for `input` element. Please check the `outlineStyles` function.

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:70]: This was is coming from the current use-cases we have in the Dashboard (see #931). But now I see it feel like over-customization, nothing will really change if we switch to one default gap size in the Dashboard. Will remove this prop.

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:62]: Good point, agree. Will make it horizontal by default (as it's the most common use case), remove this prop and make sure it can be customized with `css`.

File: packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/SelectorGroup/SelectorGroup.tsx:90]: Yep, reads much better, thank you!

File: .changeset/hot-cheetahs-press.md
Comment (mykolaharmash) [reply .changeset/hot-cheetahs-press.md:10]: Thanks, did not know about the one paragraph thing, will change.

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (connor-baer) [reply packages/circuit-ui/components/Selector/Selector.tsx:240]: I agree that the new code is more readable, however, it negatively impacts the end-user experience. 

When an app is server-side rendered, the focus state won't work until the JavaScript has fully loaded. Plus, the JavaScript focus handling could lag if the main thread is blocked by another task.

Whenever possible we should favor CSS over JavaScript, especially for UI state and micro-interactions like these. User experience trumps developer experience.

Please revert this change. I'd suggest adding a comment instead to improve the readability of the code.

Comment (mykolaharmash) [reply packages/circuit-ui/components/Selector/Selector.tsx:240]: Hm... I'm not sure I agree here. 

Selector's click handler requires JS to be loaded anyway and having a functional focus without functional click for this component does not make sense. Also, user-events (including keyboard) and layout composition happens on the main thread, so if JS is blocking it, all CSS handling will be blocked as well.

Having this said, I still respect the fact that you are the maintainer and if you think you'd be more comfortable with the CSS-based implementation in your project, I'm happy to revert this change.

Comment (connor-baer) [reply packages/circuit-ui/components/Selector/Selector.tsx:240]: > having a functional focus without functional click for this component does not make sense

There are cases where it does make sense. Users who navigate using a keyboard should be able to tab through the interface without activating the selectors.

> Also, user-events (including keyboard) and layout composition happens on the main thread, so if JS is blocking it, all CSS handling will be blocked as well. 

Fair point. 

I still would be more comfortable with the CSS-based implementation and appreciate the offer to change it back. Thank you.

**Comment by mykolaharmash** [comment]
I've addressed the latest comments. Could you guys please merge current `main` into `next`? Because `theme.borderRadius.tera` is missing in `next` right now and it fails the build.

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (connor-baer) [inline packages/circuit-ui/components/Selector/Selector.tsx:84]: 👀

**Review by connor-baer — APPROVED** [review]

**Comment by connor-baer** [comment]
@mykolaharmash The names for the border-radius values have been shifted, `tera` is now `byte` (see https://github.com/sumup-oss/circuit-ui/pull/980).

File: packages/circuit-ui/components/Selector/Selector.tsx
Comment (mykolaharmash) [reply packages/circuit-ui/components/Selector/Selector.tsx:84]: oh 🤦‍♂️

**Comment by mykolaharmash** [comment]
Fixed `borderRadius`. @connor-baer @robinmetral  feel free to merge if everything look fine 🙂

**Pull request merged**
